src/modularity.jl

```diff
@@ -1,73 +1,128 @@
 """
-    randompartition(hg::Hypergraph, n::Int)::Vector{Vector{Int}}
 
-Generates a random partition for vertices of a hypergraph `hg` into `n` subsets.
+    randompartition(h::Hypergraph, n::Int)::Vector{Set{Int}}
+
+Generates a random partition for vertices of a hypergraph `h` into `n` subsets.
 """
-randompartition(hg::Hypergraph, n::Int) = randompartition(size(hg, 1), n)
+randompartition(h::Hypergraph, n::Int) = randompartition(nhv(h), n)
 
 """
-    randompartition(hg::Hypergraph, n::Int)::Vector{Vector{Int}}
+    randompartition(h::Hypergraph, n::Int)::Vector{Set{Int}}
 
 Generates a random partition for graph having `N` vertices into `n` subsets.
 """
 function randompartition(N::Int, n::Int)
-    res = [Int[] for i in 1:n]
+    res = [Set{Int}() for i in 1:n]
     for i in 1:N
         push!(res[rand(1:n)], i)
     end
     res
 end
 
 """
-    modularity(hg::Hypergraph, partition::Vector{Vector{Int}})
 
-Calculates the modularity of a hypergraph `hg` for a given `partition`.
+HypergraphAggs(h::Hypergraph)
+
+    Precomputes vertex and edge basic stats for a hypergraph.
+    The stats are being used for efficiency reasons by community search algorithms.
 """
-@inline function modularity(hg::Hypergraph, partition::Vector{Vector{Int}})
-    @boundscheck sum(length.(partition)) == size(hg, 1)
-    @boundscheck sort!(union(partition...)) == axes(hg, 1)
-    hes = [length(hg.he2v[i]) for i in axes(hg, 2)]
-    max_hes = maximum(hes)
-    Ed = zeros(Int, max_hes)
-    for ii in 1:length(hes)
-        if hes[ii] >= 2   #no single vertex hyperedges
-            Ed[hes[ii]]+=1
+struct HypergraphAggs
+    hes::Vector{Int}
+    max_hes::Int
+    Ed::Vector{Int}
+    deg_vs::Vector{Int}
+    volV::Int    
+    HypergraphAggs(h::Hypergraph) = begin
+        hes = [length(h.he2v[i]) for i in 1:nhe(h)]
+        max_hes = maximum(hes)
+        Ed = zeros(Int, max_hes)
+        for ii in 1:length(hes)
+            if hes[ii] >= 1
+                Ed[hes[ii]]+=1
+            end
         end
+        deg_vs = [length(h.v2he[i]) for i in 1:nhv(h)]
+        volV = sum(deg_vs)
+        new(hes,max_hes,Ed,deg_vs,volV)
     end
-    deg_vs = [length(hg.v2he[i]) for i in axes(hg, 1)]
-    volV = 2*size(hg,2)
-    volP_volV = [sum(deg_vs[i] for i in p)/volV for p in partition]
-    eP = [count(i-> hes[i]>0 && (keys(hg.he2v[i]) ⊆ p), axes(hg,2)) for p in partition]
-    (sum(eP) - sum( Ed[d]*sum(volP_volV.^d) for d in 2:max_hes)) / size(hg, 2)
 end
 
+"""
+    modularity(h::Hypergraph, ha::HypergraphAggs, partition::Vector{Set{Int}})
```

Comment:
maybe make `ha` a keyword argument and give a default calling the struct constructor?